In [40]:
# Import library
import pandas as pd
import plotly.graph_objects as go
from scipy.stats import gaussian_kde
import numpy as np
import psycopg2

### Loading data

In [ ]:


# Database connection parameters
host = "34.129.179.229"
database = "postgres"
user = "postgres"
password = "Iti<FVBpR:05tTOU"
port = "5432"

# SQL query to fetch data
query = "SELECT * FROM heat_data"

# Connect to PostgreSQL and fetch data
try:
    # Establish the connection
    conn = psycopg2.connect(
        host=host,
        database=database,
        user=user,
        password=password,
        port=port
    )
    
    # Load data into a DataFrame
    df = pd.read_sql_query(query, conn)
    
    # Close the connection
    conn.close()
    
    print("Data downloaded successfully!")
except Exception as e:
    print(f"An error occurred: {e}")

/var/folders/8v/_f0s3ppn27g2ss5m_lvb311m0000gn/T/ipykernel_26337/2467433803.py:23: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Data downloaded successfully!


In [35]:
# SQL query to fetch data
query = "SELECT * FROM forest_trend"

# Connect to PostgreSQL and fetch data
try:
    # Establish the connection
    conn = psycopg2.connect(
        host=host,
        database=database,
        user=user,
        password=password,
        port=port
    )
    
    # Load data into a DataFrame
    df2 = pd.read_sql_query(query, conn)
    
    # Close the connection
    conn.close()
    
    print("Data downloaded successfully!")
except Exception as e:
    print(f"An error occurred: {e}")

Data downloaded successfully!


/var/folders/8v/_f0s3ppn27g2ss5m_lvb311m0000gn/T/ipykernel_26337/2839383081.py:16: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



In [34]:
# SQL query to fetch data
query = "SELECT * FROM global_climate_data"

# Connect to PostgreSQL and fetch data
try:
    # Establish the connection
    conn = psycopg2.connect(
        host=host,
        database=database,
        user=user,
        password=password,
        port=port
    )
    
    # Load data into a DataFrame
    df3 = pd.read_sql_query(query, conn)
    
    # Close the connection
    conn.close()
    
    print("Data downloaded successfully!")
except Exception as e:
    print(f"An error occurred: {e}")

Data downloaded successfully!


/var/folders/8v/_f0s3ppn27g2ss5m_lvb311m0000gn/T/ipykernel_26337/47942675.py:16: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



-----

### Global Climate

In [44]:
global_climate = df3.copy()
global_climate

year  avg_temp  lower_bound  upper_bound           co2       methane  \
0    1850 -0.055414    -0.226959     0.116131  1.968476e+08  1.239410e+09   
1    1851  0.128948    -0.049570     0.307466  1.987785e+08  1.244053e+09   
2    1852  0.132899    -0.047085     0.312882  2.075951e+08  1.250690e+09   
3    1853  0.091943    -0.067712     0.251598  2.172370e+08  1.258188e+09   
4    1854  0.070777    -0.070414     0.211968  2.550386e+08  1.270684e+09   
..    ...       ...          ...          ...           ...           ...   
168  2018  1.124952     1.093350     1.156554  3.673042e+10  1.019893e+10   
169  2019  1.253370     1.219085     1.287656  3.710428e+10  1.027905e+10   
170  2020  1.285218     1.250596     1.319841  3.512653e+10  1.016395e+10   
171  2021  1.124203     1.087755     1.160651  3.699173e+10  1.033092e+10   
172  2022  1.163603     1.127555     1.199651  3.729383e+10  1.052981e+10   

     ghg_emission  
0    4.219061e+09  
1    4.409311e+09  
2    4.465078e+09  
3    4.532813e+09  
4    4.594584e+09  
..            ...  
168  5.242027e+10  
169  5.279652e+10  
170  5.079254e+10  
171  5.293879e+10  
172  5.332722e+10  

[173 rows x 7 columns]

In [54]:
temp_data = pd.read_csv("https://ourworldindata.org/grapher/temperature-anomaly.csv?v=1&csvType=full&useColumnShortNames=true", storage_options = {'User-Agent': 'Our World In Data data fetch/1.0'})
world_temp_data = temp_data[temp_data['Entity'] == 'World']
world_temp_data.rename(columns={
    "near_surface_temperature_anomaly": "avg_temperature",
    "near_surface_temperature_anomaly_lower": "lower_bound",
    "near_surface_temperature_anomaly_upper": "upper_bound"
}, inplace=True)

/var/folders/8v/_f0s3ppn27g2ss5m_lvb311m0000gn/T/ipykernel_26337/3926211420.py:3: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



In [68]:
fig = go.Figure(layout=go.Layout(template='plotly_white'))

# Add traces for each data series

fig.add_trace(go.Scatter(
    x=world_temp_data['Year'],
    y=world_temp_data['upper_bound'],
    mode='lines',
    name='Upper Bound',
    line=dict(width=0),
    showlegend=False
))

fig.add_trace(go.Scatter(
    x=world_temp_data['Year'],
    y=world_temp_data['lower_bound'],
    mode='lines',
    line=dict(width=0),
    fillcolor = 'rgba(68, 68, 68, 0.3)',
    fill = 'tonexty',
    name='Lower Bound',
    showlegend=False
))

fig.add_trace(go.Scatter(
    x=world_temp_data['Year'],
    y=world_temp_data['avg_temperature'],
    mode='lines',
    name='World Average Temperature',
    line=dict(color='red')
))

fig.update_layout(
    title='Average Temperature with Confidence Interval',
    xaxis_title='Year',
    yaxis_title='Temperature (Celcius)',
    hovermode='x unified', # Shows hover info for all traces at a given x-value
    legend_title_text='Legend'
)

Line chart tracks the global average temperature anomaly (the deviation from a baseline average, shown in degrees Celsius) from the year 1850 up to the early 21st century. The red line represents the estimated world average temperature for each year, while the shaded area around the line indicates the 95% confidence interval for that year.

The chart clearly shows year-to-year fluctuations, but the dominant feature is a distinct upward trend, particularly after roughly 1950. This indicates a significant increase in the world's average temperature over time, which is a key indicator of global warming. The confidence interval narrows over time, reflecting improved accuracy in temperature measurements as technology advances. The chart effectively communicates the long-term trend of rising temperatures.

In [69]:

# Create the figure
fig = go.Figure(layout=go.Layout(template='plotly_white'))

# Add traces for each data series
fig.add_trace(go.Scatter(
    x=global_climate['year'], 
    y=global_climate['avg_temp'], 
    mode='lines', 
    name='Avg Temperature (°C)', 
    line=dict(color='red')
))

fig.add_trace(go.Scatter(
    x=global_climate['year'], 
    y=global_climate['co2'] / 1e10, 
    mode='lines', 
    name='CO₂ Emissions (x10⁻¹⁰)', 
    line=dict(color='blue')
))

fig.add_trace(go.Scatter(
    x=global_climate['year'], 
    y=global_climate['methane'] / 1e10, 
    mode='lines', 
    name='Methane Emissions (x10⁻¹⁰)', 
    line=dict(color='orange')
))

fig.add_trace(go.Scatter(
    x=global_climate['year'], 
    y=global_climate['ghg_emission'] / 1e10, 
    mode='lines', 
    name='GHG Emissions (x10⁻¹⁰)', 
    line=dict(color='green')
))

# Customize layout
fig.update_layout(
    title='Temperature and Emissions Over Time',
    xaxis_title='Year',
    yaxis_title='Values (Normalized)',
    hovermode='x unified',
    legend_title_text='Legend',
)

# Show the plot
fig.show()

Chart plots four different variables over time, from 1850 to the early 21st century:

- Average Global Temperature (Red Line)
- CO₂ Emissions (Blue Line)
- Methane Emissions (Orange Line)
- Total Greenhouse Gas (GHG) Emissions (Green Line)

The vertical axis shows normalized values, meaning the data for each line has been scaled to fit on a common axis to compare their trends rather than their absolute amounts (though the legend notes the original scale for emissions).

The key takeaway is that all four metrics show a clear increasing trend over the period, particularly accelerating from the mid-20th century (around 1950) onwards. The chart visually suggests a strong correlation between the rise in greenhouse gas emissions (CO₂, Methane, and total GHGs) and the increase in average global temperature.

---

Carbon Emission by sector

In [71]:
carbon_emission_by_sector = pd.read_csv("https://ourworldindata.org/grapher/co-emissions-by-sector.csv?v=1&csvType=full&useColumnShortNames=true", storage_options = {'User-Agent': 'Our World In Data data fetch/1.0'})
carbon_emission_by_sector.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6560 entries, 0 to 6559
Data columns (total 12 columns):
 #   Column                                        Non-Null Count  Dtype  
---  ------                                        --------------  -----  
 0   Entity                                        6560 non-null   object 
 1   Code                                          6208 non-null   object 
 2   Year                                          6560 non-null   int64  
 3   buildings_co2_emissions                       6421 non-null   float64
 4   industry_co2_emissions                        6305 non-null   float64
 5   land_use_change_and_forestry_co2_emissions    6557 non-null   float64
 6   other_fuel_combustion_co2_emissions           6421 non-null   float64
 7   transport_co2_emissions                       6421 non-null   float64
 8   manufacturing_and_construction_co2_emissions  6421 non-null   float64
 9   fugitive_co2_emissions                        2270 non-null   f

In [92]:
aus_co2_by_sector = carbon_emission_by_sector[carbon_emission_by_sector['Entity'] == 'Australia']
aus_co2 = aus_co2_by_sector[aus_co2_by_sector['Year'] == 2021]

In [ ]:
# Sum all sectors into total
aus_co2['Total'] = aus_co2.iloc[:, 3:].sum(axis=1)

# Calculate percentage of each sector
sector_percentages = aus_co2.iloc[:, 3:-1].div(aus_co2['Total'], axis=0) * 100

sector_percentages

In [ ]:
aus_co2['Total'] = aus_co2.iloc[:, 3:].sum(axis=1)



/var/folders/8v/_f0s3ppn27g2ss5m_lvb311m0000gn/T/ipykernel_26337/514574645.py:1: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



In [94]:
# calculate percentage of each sector
sector_percentages = aus_co2.iloc[:, 1:].div(aus_co2['Total'], axis=0) * 100
sector_percentages

buildings_co2_emissions  industry_co2_emissions  \
351                 3.666479                0.777251   

     land_use_change_and_forestry_co2_emissions  \
351                                    0.678084   

     other_fuel_combustion_co2_emissions  transport_co2_emissions  \
351                             2.071775                23.481011   

     manufacturing_and_construction_co2_emissions  fugitive_co2_emissions  \
351                                     10.768942                0.436868   

     electricity_and_heat_co2_emissions  aviation_and_shipping_co2_emissions  \
351                             56.7527                             1.366889   

     Total  
351  100.0

In [ ]:
sector_percentages.info()
secotr_percentages.rename(colums={
    "buildings_co2_emissions":"Buildings",
    "industry_co2_emissions":"Industry",
    "land_use_change_and_forestry_co2_emissions":"Land-use",
    "other_fuel_combustion_co2_emissions":"Other",
    transport_co2_emissions,
    manufacturing_and_construction_co2_emissions
    fugitive_co2_emissions
    electricity_and_heat_co2_emissions
    aviation_and_shipping_co2_emissions,
})

<class 'pandas.core.frame.DataFrame'>
Index: 1 entries, 351 to 351
Data columns (total 10 columns):
 #   Column                                        Non-Null Count  Dtype  
---  ------                                        --------------  -----  
 0   buildings_co2_emissions                       1 non-null      float64
 1   industry_co2_emissions                        1 non-null      float64
 2   land_use_change_and_forestry_co2_emissions    1 non-null      float64
 3   other_fuel_combustion_co2_emissions           1 non-null      float64
 4   transport_co2_emissions                       1 non-null      float64
 5   manufacturing_and_construction_co2_emissions  1 non-null      float64
 6   fugitive_co2_emissions                        1 non-null      float64
 7   electricity_and_heat_co2_emissions            1 non-null      float64
 8   aviation_and_shipping_co2_emissions           1 non-null      float64
 9   Total                                         1 non-null      float64


-----

### Forest Trend

In [30]:
forest_trend = df2.copy()
forest_trend.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 248 entries, 0 to 247
Data columns (total 4 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   country name            248 non-null    object 
 1   country code            248 non-null    object 
 2   percentage_change       248 non-null    float64
 3   percentage_change_norm  248 non-null    float64
dtypes: float64(2), object(2)
memory usage: 7.9+ KB


In [39]:
kde = gaussian_kde(forest_trend['percentage_change_norm'])
x_min = forest_trend['percentage_change_norm'].min()
x_max = forest_trend['percentage_change_norm'].max()
x_range = np.linspace(x_min - (x_max-x_min)*0.1, x_max + (x_max-x_min)*0.1, 500) # Evaluate at 500 points
y = kde(x_range)

NameError: name 'np' is not defined

----

### Heat Data

In [14]:
heat_data = df.copy()
heat_data.info()
heat_data

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8036 entries, 0 to 8035
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   max_temp  8032 non-null   float64
 1   date      8036 non-null   object 
 2   city      8036 non-null   object 
 3   is_hot    8036 non-null   bool   
dtypes: bool(1), float64(1), object(2)
memory usage: 196.3+ KB


max_temp        date       city  is_hot
0         26.5  2014-01-01  Melbourne   False
1         21.8  2014-01-02  Melbourne   False
2         21.6  2014-01-03  Melbourne   False
3         20.8  2014-01-04  Melbourne   False
4         26.2  2014-01-05  Melbourne   False
...        ...         ...        ...     ...
8031      39.5  2024-12-27     Sydney    True
8032      27.1  2024-12-28     Sydney   False
8033      29.8  2024-12-29     Sydney   False
8034      27.7  2024-12-30     Sydney   False
8035      29.6  2024-12-31     Sydney   False

[8036 rows x 4 columns]

In [24]:
heat_data['date'] = pd.to_datetime(heat_data['date'])
heat_data['year'] = heat_data['date'].dt.year

# Count hot day
hot_day = heat_data[heat_data['is_hot']]
hot_days_per_year = hot_day.groupby(['city', 'year']).size().reset_index(name='hot_day_count')

In [26]:
fig = go.Figure()

# Add traces for each city
for city in hot_days['city'].unique():
    city_data = hot_days_per_year[hot_days_per_year['city'] == city]
    fig.add_trace(go.Scatter(
        x=city_data['year'],
        y=city_data['hot_day_count'],
        mode='lines+markers',
        name=city
    ))

# Update layout
fig.update_layout(
    title='Hot Days (>30°C) Per Year',
    xaxis_title='Year',
    yaxis_title='Number of Hot Days',
    legend_title='City',
    template='plotly_white'
)

fig.show()